src/main.rs

```diff
@@ -1,3 +1,196 @@
+                        println!("Waiting for Guest Agent socket ...");
+                        thread::sleep(time::Duration::from_secs(1));
+                    }
+                    _ => break Err(err),
+                },
+            };
+        }
+    }
+
+    fn get_new_msg_id(&mut self) -> u64 {
+        self.last_msg_id += 1;
+        self.last_msg_id
+    }
+
+    fn create_header(msg_id: u64, typ: MsgType) -> Vec<u8> {
+        let mut v = Vec::new();
+        v.extend_from_slice(&msg_id.to_le_bytes());
+        v.extend_from_slice(&(typ as u8).to_le_bytes());
+        v
+    }
+
+    fn append_submsg_type(buf: &mut Vec<u8>, subtype: MsgSubType) -> () {
+        buf.extend(&(subtype as u8).to_le_bytes());
+    }
+
+    fn append_submsg_u32(buf: &mut Vec<u8>, subtype: MsgSubType, val: u32) -> () {
+        GuestAgent::append_submsg_type(buf, subtype);
+        buf.extend_from_slice(&val.to_le_bytes());
+    }
+
+    fn append_bytes(buf: &mut Vec<u8>, bytes: &[u8]) -> () {
+        buf.extend_from_slice(&bytes.len().to_le_bytes());
+        buf.extend_from_slice(bytes);
+    }
+
+    fn append_submsg_bytes(buf: &mut Vec<u8>, subtype: MsgSubType, bytes: &[u8]) -> () {
+        GuestAgent::append_submsg_type(buf, subtype);
+        GuestAgent::append_bytes(buf, bytes);
+    }
+
+    fn append_submsg_array(buf: &mut Vec<u8>, subtype: MsgSubType, array: &[&[u8]]) -> () {
+        GuestAgent::append_submsg_type(buf, subtype);
+        buf.extend_from_slice(&array.len().to_le_bytes());
+        for &a in array {
+            GuestAgent::append_bytes(buf, &a);
+        }
+    }
+
+    fn get_response(&mut self) -> Result<(), io::Error> {
+        let mut buf = [0; 8];
+        self.stream.read_exact(&mut buf)?;
+        let id = u64::from_le_bytes(buf);
+        println!("Message ID: {}", id);
+
+        let mut buf = [0; 1];
+        self.stream.read_exact(&mut buf)?;
+        let typ = RespType::from_u8(buf[0]).expect("Invalid response type!");
+        println!("Message type: {:?}", typ);
+        match typ {
+            RespType::RespOk => (),
+            RespType::RespErr => {
+                let mut buf = [0; 4];
+                self.stream.read_exact(&mut buf)?;
+                let code = u32::from_le_bytes(buf);
+                println!("Error code: {}", code);
+            }
+            RespType::NotifyOutputAvailable => (),
+            RespType::NotifyProcessDied => (),
+        };
+        Ok(())
+    }
+
+    fn quit(&mut self) -> Result<(), io::Error> {
+        let mut buf = GuestAgent::create_header(self.get_new_msg_id(), MsgType::MsgQuit);
+        GuestAgent::append_submsg_type(&mut buf, MsgSubType::SubMsgEnd);
+        self.stream.write_all(&buf)?;
+        self.get_response().unwrap();
+        Ok(())
+    }
+
+    fn run_process(
+        &mut self,
+        bin: &[u8],
+        argv: &[&[u8]],
+        maybe_env: Option<&[&[u8]]>,
+        uid: u32,
+    ) -> Result<(), io::Error> {
```

Comment:
Is there a potential case where these params aren't built from `str`s?

Reply:
None that I know of. Changed types to `&str`.